schema_utils.py

```diff
@@ -387,3 +603,88 @@ def is_valid_email(obj):
             bool. Whether the given object is a valid email.
         """
         return bool(re.search(r'^[\w\.\+\-]+\@[\w]+\.[a-z]{2,3}$', obj))
+
+    @staticmethod
+    def is_valid_asciimath_expression(obj, algebraic=True):
+        """Checks if the given  obj(a string) represents a valid algebraic or
+        numeric expression. The expression should be in the ASCIIMath format.
+        More info: http://asciimath.org/
+
+        Args:
+            obj: str. The given expression.
+            algebraic: bool. True if the given expression is algebraic
+                else numeric.
+
+        Returns:
+            bool. Whether the given object is a valid expression.
+        """
+
+        # Expression should not contain any invalid characters.
+        for character in obj:
+            if not bool(re.match(
+                    r'(\s|\d|\w|\.|\(|\)|\{|\}|\[|\]|\-|\+|\*|\/|\^)',
+                    character)):
+                return False
+
+        # Algebraic expressions should contain at least one latin letter and
+        # numeric expressions should contain none.
+        if algebraic:
+            if not bool(re.search(r'[a-zA-Z]+', obj)):
+                return False
+        else:
+            if bool(re.search(r'[a-zA-Z]+', obj)):
+                return False
+
+        # Expression should contain a balanced bracket sequence.
+        if not contains_balanced_brackets(obj):
+            return False
+
+        # Expression should be syntactically valid.
+        postfix_expression = infix_to_postfix(obj)
+        return is_valid_postfix_expression(postfix_expression)
+
+
+    @staticmethod
+    def contains_valid_latin_letters(obj):
+        """Returns True iff all elements of the given object (a list) are
+        valid latin letters.
+
+        Args:
+            obj: list(*). A list of strings.
+
+        Returns:
+            bool. Whether the given object contains valid latin letters.
+        """
+        return all(IS_ALGEBRAIC_OPERAND(elem) for elem in obj)
+
+    @staticmethod
+    def is_valid_math_equation(obj):
+        """Checks if the given  obj(a string) represents a valid math equation.
+        The expression should be in the ASCIIMath format.
+        More info: http://asciimath.org/
+
+        Args:
+            obj: str. A string.
+
+        Returns:
+            bool. Whether the given object is a valid math equation.
+        """
+        if obj.count('=') != 1:
+            return False
+
+        is_valid_asciimath_expression = get_validator(
+            'is_valid_asciimath_expression')
+        lhs, rhs = obj.split('=')
+
+        # Both sides have to be valid expressions and at least one of them has
+        # to be a valid algebraic expression.
+        if is_valid_asciimath_expression(lhs) and is_valid_asciimath_expression(
```

Comment:
I think you can save the intermediate results as variables to avoid repeated computation of the same thing.

Reply:
Done.